scripts/src/java/org/oppia/android/scripts/ci/ComputeChangedFiles.kt

```diff
@@ -167,6 +184,44 @@ class ComputeChangedFiles(
     }
   }
 
+  private fun mapTestFileToSourceFile(
+    rootDirectory: File,
+    repoRoot: String,
+    filePath: String
+  ): String? {
+    val repoRootFile = File(repoRoot).absoluteFile
+
+    val possibleSourceFilePaths = when {
+      filePath.startsWith("scripts/") -> {
+        listOf(filePath.replace("/javatests/", "/java/").replace("Test.kt", ".kt"))
+      }
+      filePath.startsWith("app/") -> {
+        when {
+          filePath.contains("/sharedTest/") -> {
+            listOf(filePath.replace("/sharedTest/", "/main/").replace("Test.kt", ".kt"))
+          }
+          filePath.contains("/test/") -> {
+            listOf(
+              filePath.replace("/test/", "/main/").replace("Test.kt", ".kt"),
+              filePath.replace("/test/", "/main/").replace("LocalTest.kt", ".kt")
+            )
+          }
+          else -> emptyList()
+        }
+      }
+      else -> {
+        listOf(filePath.replace("/test/", "/main/").replace("Test.kt", ".kt"))
+      }
```

Comment:
Are all of these different cases covered in tests? Please add tests to cover these since a regression here could be quite problematic (we'd drop re-checking code coverage for those files, leading to a potential eventual failure in the develop branch once the PR is merged).

Reply:
It did not, added the following test cases now:
- testUtility_featureBranch_appTestChanges_mapsToRespectiveSourceFile
- testUtility_featureBranch_appLocalTestChanges_mapsToRespectiveSourceFile
- testUtility_featureBranch_appSharedTestChanges_mapsToRespectiveSourceFile
- testUtility_featureBranch_scriptTestChanges_mapsToRespectiveSourceFile

And for other modules,
- testUtility_featureBranch_domainModuleTestChanges_mapsToRespectiveSourceFile
- testUtility_featureBranch_dataModuleTestChanges_mapsToRespectiveSourceFile
- testUtility_featureBranch_utilityModuleTestChanges_mapsToRespectiveSourceFile
- testUtility_featureBranch_testingModuleTestChanges_mapsToRespectiveSourceFile

---

If possible, can I understand what the following specifically means,
> to cover these since a regression here could be quite problematic (we'd drop re-checking code coverage for those files, leading to a potential eventual failure in the develop branch once the PR is merged).

Reply:
Ah thanks--these are great tests to have!

What I meant by my earlier message is essentially the following scenario:
- Other developers change this script (or utilities it depends on in the future) causing it to no longer map from test to source file correctly for, say, local tests in app module. Because there weren't unit tests, everyone fails to notice the regression and it gets checked in.
- A developer in the future performs a refactor on one of the shared tests and inadvertently reduces code coverage for the source file corresponding to that test.
- Due to the regression in this script, code coverage analysis isn't re-run for that refactored shared test and everyone misses that the code coverage dropped.
- If the coverage dropped below the enforcement threshold, then once the PR gets merged in the develop CI will start failing (since it always runs code coverage on all source files).
- The recourse here is to revert the PR that was merged and fix the code coverage gap that (eventually) caused the develop branch CI to fail.

These tests help prevent this scenario from happening.

Reply:
Ahh got it! Thank you so much for taking time and explaining it.